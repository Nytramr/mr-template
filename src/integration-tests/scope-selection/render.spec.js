import { Template } from '../../index';

describe('Template with scope-selection', () => {
  test('simple scope selection', () => {
    const scopeSelectionTpl =
      'Beginning of the file' +
      '\n' +
      'Global value is {{.value}}' +
      '\n' +
      "{{with .scope}}The scope's value is {{.value}}{{end}}" +
      '\n' +
      'End of the file' +
      '\n';

    const template = new Template('My template');
    expect(() => template.parse(scopeSelectionTpl)).not.toThrow();

    expect(
      template.execute({ value: 100, scope: { value: 200 } }),
    ).toMatchSnapshot();
    expect(
      template.execute({ value: 'a', scope: { value: 'b' } }),
    ).toMatchSnapshot();
  });

  test('scope selection with "else" fallback', () => {
    const scopeSelectionTpl =
      'Beginning of the file' +
      '\n' +
      'Global value is {{.value}}' +
      '\n' +
      '{{with .scope}}' +
      '\n' +
      "The scope's value is {{.value}}" +
      '\n' +
      '{{else}}' +
      '\n' +
      'The scope is undefined' +
      '\n' +
      '{{end}}' +
      '\n' +
      'End of the file' +
      '\n';

    const template = new Template('My template');
    expect(() => template.parse(scopeSelectionTpl)).not.toThrow();

    expect(
      template.execute({ value: 100, scope: { value: 200 } }),
    ).toMatchSnapshot();
    expect(template.execute({ value: 100 })).toMatchSnapshot();
  });

  test('scope selection with "else" and "else with" fallback', () => {
    const scopeSelectionTpl =
      'Beginning of the file' +
      '\n' +
      'Global value is {{.value}}' +
      '\n' +
      '{{with .scope1}}' +
      '\n' +
      "The scope1's value is {{.value}}" +
      '\n' +
      '{{else with .scope2}}' +
      '\n' +
      "The scope'2 value is {{.value}}" +
      '\n' +
      '{{else}}' +
      '\n' +
      'Every scope is undefined' +
      '\n' +
      '{{end}}' +
      '\n' +
      'End of the file' +
      '\n';

    const template = new Template('My template');
    expect(() => template.parse(scopeSelectionTpl)).not.toThrow();

    expect(
      template.execute({
        value: 100,
        scope1: { value: 200 },
        scope2: { value: 300 },
      }),
    ).toMatchSnapshot();
    expect(
      template.execute({
        value: 100,
        scope2: { value: 300 },
      }),
    ).toMatchSnapshot();
    expect(template.execute({ value: 100 })).toMatchSnapshot();
  });

  test('scope selection with nested scopes', () => {
    const scopeSelectionTpl = `Beginning of the file
Global value is {{.value}}
{{with .scope1}}
The scope1's value is {{.value}}
{{with ._1}}
The scope1_1's value is {{.value}}
{{else}}
The scope1_1 is undefined
{{end}}{{else with .scope2}}
The scope'2 value is {{.value}}
{{else}}
Every scope is undefined
{{end}}
End of the file`;

    const template = new Template('My template');
    expect(() => template.parse(scopeSelectionTpl)).not.toThrow();

    expect(
      template.execute({
        value: 100,
        scope1: { value: 200, _1: { value: 250 } },
        scope2: { value: 300, _1: { value: 350 } },
      }),
    ).toMatchSnapshot();
    expect(
      template.execute({
        value: 100,
        scope1: { value: 200 },
        scope2: { value: 300, _1: { value: 350 } },
      }),
    ).toMatchSnapshot();
    expect(
      template.execute({
        value: 100,
        scope2: { value: 300, _1: { value: 350 } },
      }),
    ).toMatchSnapshot();
    expect(template.execute({ value: 100 })).toMatchSnapshot();
  });
});
