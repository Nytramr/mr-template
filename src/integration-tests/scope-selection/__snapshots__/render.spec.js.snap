// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Template with scope-selection scope selection with "else" and "else with" fallback 1`] = `
"Beginning of the file
Global value is 100

The scope1's value is 200

End of the file
"
`;

exports[`Template with scope-selection scope selection with "else" and "else with" fallback 2`] = `
"Beginning of the file
Global value is 100

The scope'2 value is 300

End of the file
"
`;

exports[`Template with scope-selection scope selection with "else" and "else with" fallback 3`] = `
"Beginning of the file
Global value is 100

Every scope is undefined

End of the file
"
`;

exports[`Template with scope-selection scope selection with "else" fallback 1`] = `
"Beginning of the file
Global value is 100

The scope's value is 200

End of the file
"
`;

exports[`Template with scope-selection scope selection with "else" fallback 2`] = `
"Beginning of the file
Global value is 100

The scope is undefined

End of the file
"
`;

exports[`Template with scope-selection scope selection with nested scopes 1`] = `
"Beginning of the file
Global value is 100

The scope1's value is 200

The scope1_1's value is 250

End of the file"
`;

exports[`Template with scope-selection scope selection with nested scopes 2`] = `
"Beginning of the file
Global value is 100

The scope1's value is 200

The scope1_1 is undefined

End of the file"
`;

exports[`Template with scope-selection scope selection with nested scopes 3`] = `
"Beginning of the file
Global value is 100

The scope'2 value is 300

End of the file"
`;

exports[`Template with scope-selection scope selection with nested scopes 4`] = `
"Beginning of the file
Global value is 100

Every scope is undefined

End of the file"
`;

exports[`Template with scope-selection simple scope selection 1`] = `
"Beginning of the file
Global value is 100
The scope's value is 200
End of the file
"
`;

exports[`Template with scope-selection simple scope selection 2`] = `
"Beginning of the file
Global value is a
The scope's value is b
End of the file
"
`;
